File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Malaysia.m
//
//  UIImage+Malaysia.m
//  CountriesFlags
//

#import "Countries/UIImage+Malaysia.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (Malaysia)

+ (UIImage *)flagMalaysiaWithSize:(CGSize)size {
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorRed = [UIColor colorWithRed:204.0f/255.0f
                                        green:0.0f
                                         blue:0.0f
                                        alpha:1.0f];
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:0.0f
                                          blue:102.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:204.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorRed setFill];
    for (int index = 0; index <= 12; index += 2) {
        UIRectFill(CGRectMake(0.0, size.height / 14.0f * index, size.width, size.height / 14.0f));
    }

    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width / 2.0f, size.height / 14.0f * 8.0f));

    [colorYellow setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGPointMake(size.width / 28.0 * 5.75, size.height / 14.0 * 4.0)
                                    radius:size.height / 14.0f * 3.0f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorBlue setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGPointMake(size.width / 28.0f * 6.4166f, size.height / 14.0f * 4.0f)
                                    radius:size.height / 14.0f * 2.6665f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorYellow setFill];
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 28.0f * 8.75f, size.height / 14.0f * 4.0f)
                        radius:size.height / 14.0f * 2.5f
                          type:FlagStarTypeFourteenPointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Malaysia.swift
//
//  UIImage+Malaysia.swift
//  CountriesFlags
//

// swiftlint:disable function_body_length

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Malaysia "wikipedia"
     
     Generates the flag of Malaysia.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Malaysia.
     */
    internal static func flagMalaysia(with size: CGSize) -> UIImage? {
        let colorWhite = UIColor.white
        let colorRed = UIColor(red: 204.0 / 255.0,
                               green: 0.0,
                               blue: 0.0,
                               alpha: 1.0)
        let colorBlue = UIColor(red: 0.0,
                                green: 0.0,
                                blue: 102.0 / 255.0,
                                alpha: 1.0)
        let colorYellow = UIColor(red: 1.0,
                                  green: 204.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorWhite.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorRed.setFill()
        for index in stride(from: 0.0, through: 12.0, by: 2.0) {
            UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 14.0 * index),
                              size: CGSize(width: size.width, height: size.height / 14.0)))
        }

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: CGSize(width: size.width / 2.0, height: size.height / 14.0 * 8.0)))

        colorYellow.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.width / 28.0 * 5.75, y: size.height / 14.0 * 4.0),
                     radius: size.height / 14.0 * 6.0 / 2.0,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorBlue.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.width / 28.0 * 6.4166, y: size.height / 14.0 * 4.0),
                     radius: size.height / 14.0 * 5.333 / 2.0,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorYellow.setFill()
        UIBezierPath.star(with: CGSize(width: size.width / 28.0 * 17.5, height: size.height / 14.0 * 8.0),
                          radius: size.height / 14.0 * 5.0 / 2.0,
                          type: .fourteenPointed).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif

// swiftlint:enable function_body_length
